src/mlpack/methods/ann/augmented/ham_unit_impl.hpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * @file ham_unit_impl.hpp
+ * @author Konstantin Sidorov
+ *
+ * Implementation of HAMUnit class
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+#define MLPACK_METHODS_AUGMENTED_HAM_UNIT_IMPL_HPP
+
+#include "ham_unit.hpp"
+
+namespace mlpack {
+namespace ann /* Artificial Neural Network */ {
+namespace augmented /* Augmented neural network */ {
+
+template<
+  typename C,
+  typename E,
+  typename J,
+  typename S,
+  typename W
+>
+HAMUnit::HAMUnit(int memorySize,
+                 C& controller,
+                 E& embed,
+                 J& join,
+                 S& search,
+                 W& write)
+  : memorySize(memorySize),
+    search(search), embed(embed), controller(controller)
+{
+  memory = TreeMemory<double, J, W>(memorySize, join, write);
+}
+
+template<
+  typename C,
+  typename E,
+  typename J,
+  typename S,
+  typename W
+>
+arma::vec HAMUnit::Attention() const
+{
+  size_t nodesCnt = 2 * memory.ActualMemorySize() - 1;
+  arma::vec probabilities(nodesCnt);
+  probabilities(0) = 1;
+  // TODO Query the controller state.
```

Comment:
Do you think we should start with the Raw HAM as described in 4.3 and later extend the model with the LSTM controller? Let me know what you think.